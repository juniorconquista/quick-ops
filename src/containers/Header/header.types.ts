type CostCenter = {
    id: number;
    name: string;
};

export interface HeaderResponse {
    justification: string;
    purpose: string;
    createdOn: any;
    analyst: boolean;
    project: {
        title: string;
    };
    collaborator: {
        name: string;
        email: string;
    };
    accountabilityExtraInfo: {
        amountOfPeople: number;
        eventDate: any;
        budgetForBreakfast: boolean;
    };
    costCenters: CostCenter[];
}
