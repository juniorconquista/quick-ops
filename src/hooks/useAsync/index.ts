import useAsync from './useAsync';
export * from './useAsync.utils';

export default useAsync;
