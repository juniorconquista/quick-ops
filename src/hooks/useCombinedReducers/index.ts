import useCombinedReducers from './useCombinedReducers';

export default useCombinedReducers;
