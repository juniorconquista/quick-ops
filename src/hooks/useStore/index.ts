import useStore from './useStore';
export * from './useStore';

export default useStore;
