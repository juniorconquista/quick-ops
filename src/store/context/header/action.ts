import { ISLOADING, ERROR, RESET } from './types';

export const isLoadingAction = (isLoading: boolean) => {
    return {
        type: ISLOADING,
        payload: isLoading,
    };
};

export const setErrorAction = (error: string) => {
    return {
        type: ERROR,
        payload: error,
    };
};

export const ResetAction = () => {
    return {
        type: RESET,
    };
};
