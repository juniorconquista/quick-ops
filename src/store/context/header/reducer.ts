import { ISLOADING, ERROR } from './types';

export const headerInitialState = {
    header: {
        data: {},
        isLoading: false,
        error: null,
    },
};

const headerReducer = (
    state = headerInitialState,
    action: { type: string; payload: any },
) => {
    switch (action.type) {
        case ISLOADING:
            return {
                header: {
                    ...state.header,
                    isLoading: action.payload,
                },
            };
        case ERROR:
            return {
                header: {
                    ...state.header,
                    error: action.payload,
                },
            };
        default:
            return state;
    }
};

export default headerReducer;
