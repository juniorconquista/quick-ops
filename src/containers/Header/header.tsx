import React, { memo, useEffect } from 'react';
import { Box, H1, H2 } from 'quarks-ui';
import { css } from 'styled-components';

import useAsync from '../../hooks/useAsync';
import Loading from '../../components/Loading';
import { HeaderResponse } from './header.types';
import Item from './Item';
import Select from '../../components/Select';

const Header: React.FC = () => {
    const getHeader = (): Promise<HeaderResponse> =>
        fetch(`${process.env.REACT_APP_API_URL}/header`).then((response) =>
            response.json(),
        );

    const { execute, status, value } = useAsync<HeaderResponse>(
        getHeader,
        false,
    );

    // useEffect(() => { execute() }, []);  // eslint-disable-line react-hooks/exhaustive-deps

    return (
        <Box
            styling="column"
            position="relative"
            width="100%"
            borderRadius="12px"
            minHeight="236px"
            p="24px 30px"
            justifyContent="space-between"
            alignItems="flex-start"
            css={css({
                backgroundImage:
                    'linear-gradient(to right, #00d7ce 3%, #00c5fc)',
            })}
        >
            <H1 styling="base" fontWeight="bold" fontSize="20px" color="#FFF">
                {value?.project.title ?? '-'}
            </H1>
            <Box
                styling="grid"
                width="100%"
                minHeight="150px"
                gridGap="10px"
                gridTemplateColumns="repeat(3, 1fr)"
                css={css({
                    '@media (max-width: 1200px)': {
                        gridTemplateColumns: 'repeat(1, 1fr)',
                    },
                })}
            >
                <Box
                    styling="column"
                    height="100%"
                    borderRight="solid 1px #fafbfc"
                    p="10px 0px"
                    css={css({
                        '@media (max-width: 1200px)': {
                            borderRight: 'none',
                            borderBottom: 'solid 1px #fafbfc',
                            paddingBottom: '0px',
                        },
                    })}
                >
                    <Item
                        title="Name"
                        description={value?.collaborator.name ?? '-'}
                    />
                    <Item
                        title="Email"
                        description={value?.collaborator.email ?? '-'}
                    />
                    <Item
                        title="Justificativa"
                        description={value?.justification ?? '-'}
                    />
                    <Item
                        title="Finalidade"
                        description={value?.purpose ?? '-'}
                    />
                </Box>
                <Box
                    styling="column"
                    height="100%"
                    borderRight="solid 1px #fafbfc"
                    p="10px 0px"
                    css={css({
                        '@media (max-width: 1200px)': {
                            borderRight: 'none',
                            borderBottom: 'solid 1px #fafbfc',
                            padding: '0px',
                        },
                    })}
                >
                    {value?.costCenters.map((costCenter) => (
                        <Item
                            key={costCenter.id}
                            title="Centro de custo"
                            description={costCenter?.name ?? '-'}
                        />
                    ))}
                    <Item
                        title="Projeto"
                        description={value?.project.title ?? '-'}
                    />
                    <Item
                        title="Data"
                        description={
                            value?.accountabilityExtraInfo.eventDate ?? '-'
                        }
                    />
                </Box>
                <Box
                    styling="column"
                    height="100%"
                    p="10px 0px"
                    css={css({
                        '@media (max-width: 1200px)': {
                            borderRight: 'none',
                            padding: '0px',
                        },
                    })}
                >
                    <Item
                        title="Quantiade"
                        description={`${
                            value?.accountabilityExtraInfo.amountOfPeople ?? '-'
                        }`}
                    />
                    <Item
                        title="Inclui café da manhã"
                        description={`${
                            value?.accountabilityExtraInfo.budgetForBreakfast
                                ? 'Sim'
                                : 'Não'
                        }`}
                    />
                    {value?.analyst && (
                        <Box
                            styling="column"
                            width="100%"
                            alignItems="flex-start"
                        >
                            <H2
                                styling="base"
                                fontWeight="normal"
                                fontSize="14px"
                                mr="15px"
                                textAlign="right"
                                color="#FFF"
                                mb="5px"
                            >
                                Atribuir analista
                            </H2>
                            <Select
                                styling="base"
                                placeholder="Selecione"
                                options={[]}
                                name="analyst"
                                bg="#FFF"
                            />
                        </Box>
                    )}
                </Box>
            </Box>
            {status === 'pending' && <Loading />}
        </Box>
    );
};

export default memo(Header);
