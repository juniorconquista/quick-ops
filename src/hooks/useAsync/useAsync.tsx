import { useState, useEffect, useCallback } from 'react';
import { Status } from './useAsync.types';

const useAsync = <T, E = string>(
    asyncFunction: (body?: any) => Promise<T>,
    immediate = true,
) => {
    const [status, setStatus] = useState<Status>('idle');
    const [value, setValue] = useState<T | null>(null);
    const [error, setError] = useState<E | null>(null);

    const execute = useCallback(
        (body?: any) => {
            setStatus('pending');
            setValue(null);
            setError(null);

            return asyncFunction(body)
                .then((response: any) => {
                    setValue(response);
                    setStatus('success');
                })
                .catch((error: any) => {
                    setError(error);
                    setStatus('error');
                });
        },
        [asyncFunction],
    );

    useEffect(() => {
        if (immediate) {
            execute();
        }
    }, [execute, immediate]);

    return { execute, status, value, error };
};

export default useAsync;
