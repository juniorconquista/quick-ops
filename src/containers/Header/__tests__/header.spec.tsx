import React from 'react';
import { cleanup, render } from '@testing-library/react';

import { HeaderFaker } from '../../../mocks/header.faker';
import * as useAsync from '../../../hooks/useAsync';
import ThemeProvider from '../../../utils/ThemeProviderTest';
import Header from '../header';
import 'jest-styled-components';

const setupFetchStub: any = (data: any) => (_url: string) =>
    new Promise((resolve) => {
        resolve({
            json: () =>
                Promise.resolve({
                    data,
                }),
        });
    });

const headerResponseFaker = new HeaderFaker();
const headerResponse = headerResponseFaker.generate();

describe('Header container', () => {
    let mockUseAsyncSpyFn: any;

    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });

    beforeEach(() => {
        // spyOn's
        const fakeData = { fake: 'data' };
        jest.spyOn(global, 'fetch').mockImplementation(
            setupFetchStub(fakeData),
        );

        const useAsyncSpy = jest.spyOn(useAsync, 'default');
        mockUseAsyncSpyFn = {
            ...jest.requireActual('../../../hooks/useAsync'),
            execute: jest.fn(),
            value: headerResponse,
        };
        useAsyncSpy.mockImplementation(
            jest.fn((fn) => {
                fn();
                return mockUseAsyncSpyFn;
            }),
        );
    });

    it('tests basic snapshot', () => {
        const useAsyncSpy = jest.spyOn(useAsync, 'default');
        const mockUseAsyncSpyFn = {
            ...jest.requireActual('../../../hooks/useAsync'),
            execute: jest.fn(),
            value: {},
        };
        useAsyncSpy.mockImplementation(
            jest.fn((fn) => {
                fn();
                return mockUseAsyncSpyFn;
            }),
        );

        const container = render(
            <ThemeProvider>
                <Header />
            </ThemeProvider>,
        );
        expect(container.asFragment()).toMatchSnapshot();
    });

    it('should call the api to bring the necessary data when rendering the header', () => {
        render(
            <ThemeProvider>
                <Header />
            </ThemeProvider>,
        );
        expect(mockUseAsyncSpyFn.execute).toBeCalled();
    });
});
