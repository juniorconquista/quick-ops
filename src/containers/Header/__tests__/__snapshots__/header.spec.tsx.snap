// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header container tests basic snapshot 1`] = `
<DocumentFragment>
  .c1 {
  color: #f8f9fa;
  font-size: 2rem;
  color: #FFF;
  font-weight: bold;
  font-size: 20px;
}

.c5 {
  color: #f8f9fa;
  font-size: 1.5rem;
  color: #FFF;
  font-weight: normal;
  font-size: 14px;
  text-align: right;
  margin-right: 15px;
}

.c6 {
  color: #f8f9fa;
  font-size: 1rem;
  color: #FFF;
  font-weight: bold;
  font-size: 14px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 24px 30px;
  position: relative;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  width: 100%;
  min-height: 236px;
  border-radius: 12px;
  background-image: linear-gradient(to right,#00d7ce 3%,#00c5fc);
}

.c0:focus {
  outline: none;
}

.c2 {
  display: grid;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  grid-gap: 10px;
  grid-template-columns: repeat(3,1fr);
  width: 100%;
  min-height: 150px;
}

.c2:focus {
  outline: none;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 0px;
  height: 100%;
  border-right: solid 1px #fafbfc;
}

.c3:focus {
  outline: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 10px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  width: 100%;
}

.c4:focus {
  outline: none;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 0px;
  height: 100%;
  border-right: solid 1px #fafbfc;
}

.c7:focus {
  outline: none;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 0px;
  height: 100%;
}

.c8:focus {
  outline: none;
}

@media (max-width:1200px) {
  .c2 {
    grid-template-columns: repeat(1,1fr);
  }
}

@media (max-width:1200px) {
  .c3 {
    border-right: none;
    border-bottom: solid 1px #fafbfc;
    padding-bottom: 0px;
  }
}

@media (max-width:1200px) {
  .c7 {
    border-right: none;
    border-bottom: solid 1px #fafbfc;
    padding: 0px;
  }
}

@media (max-width:1200px) {
  .c8 {
    border-right: none;
    padding: 0px;
  }
}

<div
    class="c0"
    width="100%"
  >
    <h1
      class="c1"
      color="#FFF"
      font-size="20px"
      font-weight="bold"
    >
      -
    </h1>
    <div
      class="c2"
      width="100%"
    >
      <div
        class="c3"
        height="100%"
      >
        <div
          class="c4"
          width="100%"
        >
          <h2
            class="c5"
            color="#FFF"
            font-size="14px"
            font-weight="normal"
          >
            Name
          </h2>
          <p
            class="c6"
            color="#FFF"
            font-size="14px"
            font-weight="bold"
          >
            -
          </p>
        </div>
        <div
          class="c4"
          width="100%"
        >
          <h2
            class="c5"
            color="#FFF"
            font-size="14px"
            font-weight="normal"
          >
            Email
          </h2>
          <p
            class="c6"
            color="#FFF"
            font-size="14px"
            font-weight="bold"
          >
            -
          </p>
        </div>
        <div
          class="c4"
          width="100%"
        >
          <h2
            class="c5"
            color="#FFF"
            font-size="14px"
            font-weight="normal"
          >
            Justificativa
          </h2>
          <p
            class="c6"
            color="#FFF"
            font-size="14px"
            font-weight="bold"
          >
            -
          </p>
        </div>
        <div
          class="c4"
          width="100%"
        >
          <h2
            class="c5"
            color="#FFF"
            font-size="14px"
            font-weight="normal"
          >
            Finalidade
          </h2>
          <p
            class="c6"
            color="#FFF"
            font-size="14px"
            font-weight="bold"
          >
            -
          </p>
        </div>
      </div>
      <div
        class="c7"
        height="100%"
      >
        <div
          class="c4"
          width="100%"
        >
          <h2
            class="c5"
            color="#FFF"
            font-size="14px"
            font-weight="normal"
          >
            Projeto
          </h2>
          <p
            class="c6"
            color="#FFF"
            font-size="14px"
            font-weight="bold"
          >
            -
          </p>
        </div>
        <div
          class="c4"
          width="100%"
        >
          <h2
            class="c5"
            color="#FFF"
            font-size="14px"
            font-weight="normal"
          >
            Data
          </h2>
          <p
            class="c6"
            color="#FFF"
            font-size="14px"
            font-weight="bold"
          >
            -
          </p>
        </div>
      </div>
      <div
        class="c8"
        height="100%"
      >
        <div
          class="c4"
          width="100%"
        >
          <h2
            class="c5"
            color="#FFF"
            font-size="14px"
            font-weight="normal"
          >
            Quantiade
          </h2>
          <p
            class="c6"
            color="#FFF"
            font-size="14px"
            font-weight="bold"
          >
            -
          </p>
        </div>
        <div
          class="c4"
          width="100%"
        >
          <h2
            class="c5"
            color="#FFF"
            font-size="14px"
            font-weight="normal"
          >
            Inclui café da manhã
          </h2>
          <p
            class="c6"
            color="#FFF"
            font-size="14px"
            font-weight="bold"
          >
            Não
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
